datumaro/components/project.py

```diff
@@ -2,486 +2,2399 @@
-        dataset_save_dir = osp.join(save_dir, project.config.dataset_dir)
+    def add_remote(self, name, url):
+        self.repo.create_remote(name, url)
 
-        converter_kwargs = {
-            'save_images': save_images,
-        }
+    def remove_remote(self, name):
+        self.repo.delete_remote(name)
 
-        save_dir_existed = osp.exists(save_dir)
+    def is_ref(self, rev):
         try:
-            os.makedirs(save_dir, exist_ok=True)
-            os.makedirs(dataset_save_dir, exist_ok=True)
+            self.repo.commit(rev)
+            return True
+        except (ValueError, self.module.exc.BadName):
+            return False
+
+    def has_commits(self):
+        return self.is_ref('HEAD')
+
+    def show(self, path, rev=None):
+        return self.repo.git.show('%s:%s' % (rev or '', path))
+
+    def get_tree(self, ref):
+        return self.repo.tree(ref)
+
+    def write_tree(self, tree, base_path: str,
+            include_files: Optional[List[str]] = None):
+        os.makedirs(base_path, exist_ok=True)
+
+        for obj in tree.traverse(visit_once=True):
+            if include_files and obj.path not in include_files:
+                continue
+
+            path = osp.join(base_path, obj.path)
+            os.makedirs(osp.dirname(path), exist_ok=True)
+            if obj.type == 'blob':
+                with open(path, 'wb') as f:
+                    obj.stream_data(f)
+            elif obj.type == 'tree':
+                pass
+            else:
+                raise ValueError("Unexpected object type in a "
+                    "git tree: %s (%s)" % (obj.type, obj.hexsha))
+
+    @property
+    def head(self) -> str:
+        return self.repo.head.commit.hexsha
+
+    @property
+    def branch(self) -> str:
+        if self.repo.head.is_detached:
+            return None
+        return self.repo.active_branch
+
+    def rev_parse(self, ref: str) -> Tuple[str, str]:
+        """
+        Expands named refs and tags.
+
+        Returns: object type, object hash
+        """
+        obj = self.repo.rev_parse(ref)
+        return obj.type, obj.hexsha
+
+    def ignore(self, paths: Union[str, List[str]],
+            mode: Union[None, str, IgnoreMode] = None,
+            gitignore: Optional[str] = None):
+        if not gitignore:
+            gitignore = '.gitignore'
+        repo_root = self._project_dir
+        gitignore = osp.abspath(osp.join(repo_root, gitignore))
+        assert gitignore.startswith(repo_root), gitignore
+
+        _update_ignore_file(paths, repo_root=repo_root,
+            mode=mode, filepath=gitignore)
+
+    HASH_LEN = 40
+
+    @classmethod
+    def is_hash(cls, s: str) -> bool:
+        return len(s) == cls.HASH_LEN
+
+    def log(self, depth=10) -> List[str]:
+        commits = []
+
+        if not self.has_commits():
+            return commits
+
+        for commit in zip(self.repo.iter_commits(rev='HEAD'), range(depth)):
+            commits.append(commit)
```

Comment:
Either the function declaration or the code is wrong. This puts a tuple in the resulting list, not a string.

Reply:
Fixed the header